File: backend/services/bm25.service.js
/**
 * BM25 Service — Term Frequency scoring for internship retrieval
 * Classic BM25 ranking function: k1=1.5, b=0.75
 */

const k1 = 1.5;
const b = 0.75;

function tokenize(text) {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, ' ')
    .split(/\s+/)
    .filter((t) => t.length > 1);
}

function buildCorpus(internships) {
  return internships.map((doc) => {
    const text = [doc.title, doc.description, ...(doc.skills || []), doc.location, doc.type].join(' ');
    return tokenize(text);
  });
}

function computeIDF(corpus) {
  const N = corpus.length;
  const df = {};
  corpus.forEach((doc) => {
    const uniqueTerms = new Set(doc);
    uniqueTerms.forEach((term) => {
      df[term] = (df[term] || 0) + 1;
    });
  });
  const idf = {};
  Object.entries(df).forEach(([term, count]) => {
    idf[term] = Math.log((N - count + 0.5) / (count + 0.5) + 1);
  });
  return idf;
}

function bm25Score(queryTokens, docTokens, idf, avgDocLen) {
  const docLen = docTokens.length;
  const tf = {};
  docTokens.forEach((t) => { tf[t] = (tf[t] || 0) + 1; });

  let score = 0;
  queryTokens.forEach((term) => {
    if (!idf[term]) return;
    const termFreq = tf[term] || 0;
    const numerator = termFreq * (k1 + 1);
    const denominator = termFreq + k1 * (1 - b + b * (docLen / avgDocLen));
    score += idf[term] * (numerator / denominator);
  });
  return score;
}

/**
 * Rank internships against a query using BM25
 * @param {string} query - search query
 * @param {Array} internships - array of internship documents
 * @returns {Array} internships with bm25Score, sorted descending
 */
function rankWithBM25(query, internships) {
  if (!internships || internships.length === 0) return [];

  const corpus = buildCorpus(internships);
  const idf = computeIDF(corpus);
  const avgDocLen = corpus.reduce((sum, doc) => sum + doc.length, 0) / corpus.length;
  const queryTokens = tokenize(query);

  const scored = internships.map((internship, idx) => ({
    internship,
    bm25Score: bm25Score(queryTokens, corpus[idx], idf, avgDocLen),
  }));

  return scored.sort((a, b) => b.bm25Score - a.bm25Score);
}

module.exports = { rankWithBM25, tokenize };

File: ml_backend/services/bm25_service.py
"""
BM25 Service — Okapi BM25 ranking using rank_bm25 library.
Falls back to simple TF scoring if rank_bm25 is unavailable.
"""

import re
import math
import logging
from typing import List, Dict, Any

logger = logging.getLogger(__name__)


def tokenize(text: str) -> List[str]:
    """Lowercase, strip punctuation, split on whitespace, filter short tokens."""
    return [
        t for t in re.sub(r"[^a-z0-9\s]", " ", (text or "").lower()).split()
        if len(t) > 1
    ]


def build_corpus_text(internship: Dict[str, Any]) -> str:
    """Build a single text string from internship fields for BM25 indexing."""
    return " ".join([
        internship.get("title", ""),
        internship.get("description", ""),
        " ".join(internship.get("skills", [])),
        internship.get("location", ""),
        internship.get("type", ""),
        internship.get("company", ""),
    ])


def rank_bm25(query: str, internships: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Rank internships by BM25 score against the query.

    Returns list of dicts: [{internship, bm25_score}, ...] sorted descending.
    Normalizes scores to [0, 1].
    """
    if not internships:
        return []

    corpus = [tokenize(build_corpus_text(i)) for i in internships]
    query_tokens = tokenize(query)

    # Use rank_bm25 if available
    try:
        from rank_bm25 import BM25Okapi
        bm25 = BM25Okapi(corpus, k1=1.5, b=0.75)
        raw_scores = bm25.get_scores(query_tokens).tolist()
    except ImportError:
        logger.warning("rank_bm25 not installed, using simple TF scoring")
        raw_scores = [_simple_tf_score(query_tokens, doc) for doc in corpus]

    # Normalize to [0, 1]
    max_score = max(raw_scores) if raw_scores else 1.0
    max_score = max_score if max_score > 0 else 1.0
    normalized = [s / max_score for s in raw_scores]

    results = [
        {"internship": internships[i], "bm25_score": normalized[i]}
        for i in range(len(internships))
    ]
    return sorted(results, key=lambda x: x["bm25_score"], reverse=True)


def _simple_tf_score(query_tokens: List[str], doc_tokens: List[str]) -> float:
    """Simple TF-based score as fallback when rank_bm25 is unavailable."""
    doc_set = set(doc_tokens)
    return sum(1.0 for t in query_tokens if t in doc_set)
